File: mcp-list/packages/EventAnalyzer/src/code_searcher.py
"""Code Searcher
在项目代码中搜索字段实现
"""

import os
import re
from pathlib import Path
from typing import List, Dict, Any


class CodeSearcher:
    """代码搜索器"""

    # 支持的文件扩展名
    SUPPORTED_EXTENSIONS = ('.js', '.ts', '.vue', '.jsx', '.tsx', '.py', '.java', '.php')

    # 排除的目录
    EXCLUDED_DIRS = {
        'node_modules', '.git', 'dist', 'build', '__pycache__',
        'venv', 'vendor', '.next', '.nuxt', 'coverage'
    }

    def find_field(
        self,
        field_name: str,
        project_path: str,
        max_results: int = 50
    ) -> Dict[str, Any]:
        """
        在项目中搜索字段实现

        Args:
            field_name: 字段名称
            project_path: 项目路径
            max_results: 最大结果数

        Returns:
            搜索结果
        """
        if not os.path.exists(project_path):
            return {
                "field": field_name,
                "error": f"项目路径不存在: {project_path}",
                "found_locations": [],
                "total_matches": 0
            }

        matches = []

        # 搜索模式
        patterns = self._build_search_patterns(field_name)

        # 遍历项目文件
        for file_path in self._iter_project_files(project_path):
            if len(matches) >= max_results:
                break

            file_matches = self._search_in_file(file_path, patterns, field_name)
            matches.extend(file_matches)

        return {
            "field": field_name,
            "found_locations": matches[:max_results],
            "total_matches": len(matches),
            "truncated": len(matches) > max_results
        }

    def _build_search_patterns(self, field_name: str) -> List[re.Pattern]:
        """
        构建搜索模式

        Args:
            field_name: 字段名称

        Returns:
            正则表达式模式列表
        """
        patterns = [
            # 字符串字面量: "field_name" 或 'field_name'
            re.compile(rf'["\']({field_name})["\']'),

            # 对象属性: field_name: value
            re.compile(rf'\b({field_name})\s*:'),

            # 对象属性赋值: obj.field_name = value
            re.compile(rf'\.({field_name})\s*='),

            # 变量名: const/let/var field_name
            re.compile(rf'\b(?:const|let|var)\s+({field_name})\b'),

            # 解构赋值: { field_name }
            re.compile(rf'\{{\s*({field_name})\s*\}}'),
        ]

        return patterns

    def _iter_project_files(self, project_path: str):
        """
        迭代项目中的所有支持的文件

        Args:
            project_path: 项目路径

        Yields:
            文件路径
        """
        for root, dirs, files in os.walk(project_path):
            # 排除指定目录
            dirs[:] = [d for d in dirs if d not in self.EXCLUDED_DIRS]

            for file in files:
                if file.endswith(self.SUPPORTED_EXTENSIONS):
                    yield os.path.join(root, file)

    def _search_in_file(
        self,
        file_path: str,
        patterns: List[re.Pattern],
        field_name: str
    ) -> List[Dict[str, Any]]:
        """
        在单个文件中搜索

        Args:
            file_path: 文件路径
            patterns: 搜索模式列表
            field_name: 字段名称

        Returns:
            匹配结果列表
        """
        matches = []

        try:
            with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                lines = f.readlines()

            for line_num, line in enumerate(lines, 1):
                for pattern in patterns:
                    if pattern.search(line):
                        # 获取上下文（前后各1行）
                        context_lines = []
                        for i in range(max(0, line_num - 2), min(len(lines), line_num + 1)):
                            context_lines.append(lines[i].rstrip())

                        matches.append({
                            "file": file_path,
                            "line": line_num,
                            "code": line.strip(),
                            "context": "\\n".join(context_lines)
                        })
                        break  # 每行只记录一次

        except Exception as e:
            # 忽略读取错误
            pass

        return matches

File: mcp-list/packages/EventAnalyzer-node/src/services/codeSearcher.ts
/**
 * 代码搜索器
 * 在项目代码中搜索字段的使用位置
 */

import * as fs from 'fs';
import * as path from 'path';

export interface SearchMatch {
  file: string;
  line: number;
  code: string;
  context_before: string[];
  context_after: string[];
}

export interface SearchResult {
  field_name: string;
  project_path: string;
  total_matches: number;
  matches: SearchMatch[];
  searched_files: number;
}

export class CodeSearcher {
  // 支持的文件扩展名
  private supportedExtensions = [
    '.ts',
    '.tsx',
    '.js',
    '.jsx',
    '.vue',
    '.json',
    '.py',
    '.java',
    '.kt',
    '.swift',
  ];

  // 忽略的目录
  private ignoredDirs = ['node_modules', '.git', 'dist', 'build', 'coverage', '.next', 'out'];

  /**
   * 在项目中搜索字段
   * @param fieldName 字段名称
   * @param projectPath 项目根目录
   * @param maxResults 最大结果数
   */
  findField(fieldName: string, projectPath: string, maxResults: number = 50): SearchResult {
    // 检查项目路径是否存在
    if (!fs.existsSync(projectPath)) {
      throw new Error(`项目路径不存在: ${projectPath}`);
    }

    const matches: SearchMatch[] = [];
    let searchedFiles = 0;

    // 递归搜索文件
    this.searchDirectory(projectPath, fieldName, matches, maxResults, (count) => {
      searchedFiles = count;
    });

    return {
      field_name: fieldName,
      project_path: projectPath,
      total_matches: matches.length,
      matches,
      searched_files: searchedFiles,
    };
  }

  /**
   * 递归搜索目录
   */
  private searchDirectory(
    dirPath: string,
    fieldName: string,
    matches: SearchMatch[],
    maxResults: number,
    onFileSearched: (count: number) => void
  ): void {
    if (matches.length >= maxResults) {
      return;
    }

    let fileCount = 0;

    try {
      const entries = fs.readdirSync(dirPath, { withFileTypes: true });

      for (const entry of entries) {
        if (matches.length >= maxResults) {
          break;
        }

        const fullPath = path.join(dirPath, entry.name);

        if (entry.isDirectory()) {
          // 跳过忽略的目录
          if (!this.ignoredDirs.includes(entry.name)) {
            this.searchDirectory(fullPath, fieldName, matches, maxResults, onFileSearched);
          }
        } else if (entry.isFile()) {
          // 检查文件扩展名
          const ext = path.extname(entry.name);
          if (this.supportedExtensions.includes(ext)) {
            fileCount++;
            onFileSearched(fileCount);
            this.searchFile(fullPath, fieldName, matches, maxResults);
          }
        }
      }
    } catch (error) {
      // 跳过无法访问的目录
    }
  }

  /**
   * 在单个文件中搜索
   */
  private searchFile(
    filePath: string,
    fieldName: string,
    matches: SearchMatch[],
    maxResults: number
  ): void {
    if (matches.length >= maxResults) {
      return;
    }

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const lines = content.split('\n');

      for (let i = 0; i < lines.length; i++) {
        if (matches.length >= maxResults) {
          break;
        }

        const line = lines[i];

        // 检查字段名是否出现在当前行
        if (this.containsField(line, fieldName)) {
          matches.push({
            file: filePath,
            line: i + 1,
            code: line.trim(),
            context_before: this.getContext(lines, i - 2, i),
            context_after: this.getContext(lines, i + 1, i + 3),
          });
        }
      }
    } catch (error) {
      // 跳过无法读取的文件
    }
  }

  /**
   * 检查行中是否包含字段名
   */
  private containsField(line: string, fieldName: string): boolean {
    // 使用正则表达式匹配完整的字段名（避免部分匹配）
    const patterns = [
      new RegExp(`["'\`]${fieldName}["'\`]`, 'i'), // 字符串形式："fieldName"
      new RegExp(`\\b${fieldName}\\b`, 'i'), // 作为标识符
      new RegExp(`${fieldName}:`, 'i'), // 对象属性
      new RegExp(`\\.${fieldName}\\b`, 'i'), // 属性访问
    ];

    return patterns.some((pattern) => pattern.test(line));
  }

  /**
   * 获取代码上下文
   */
  private getContext(lines: string[], start: number, end: number): string[] {
    const context: string[] = [];

    for (let i = Math.max(0, start); i < Math.min(lines.length, end); i++) {
      context.push(lines[i].trim());
    }

    return context;
  }
}
